types.go

```diff
@@ -14,9 +14,12 @@ type Service interface {
 	SyncWith(ctx context.Context, adapter Adapter) error // Sync the things in a source service with this service.
 }
 
+// A ConfigFn is used to pass additional or custom functionality to an adapter.
+type ConfigFn = func(interface{})
+
 // InitFn is an optional adapter function that can initialise a new adapter using a static configuration.
 // This is to make it easier to use an adapter in a CLI or other service that invokes adapters programmatically.
-type InitFn = func(ctx context.Context, config map[string]string) (Adapter, error)
+type InitFn = func(ctx context.Context, config map[string]string, configFns ...ConfigFn) (Adapter, error)
```

Comment:
I really like this pattern for setting configuration. I'm wondering whether there is a possibility of tiding this up with generics, to remove that ugly `interface{}`? I'm not too hot on them, but does something like this work?

```
type ConfigFn[T Adapter] func(T)

type InitFn[T Adapter] func(context.Context, map[ConfigKey]string, ...ConfigFn[T]) (T, error)
```

Reply:
I think the best we can get is 
```go
type ConfigFn = func(Adapter)
```

which isn't _terrible_. I might actually make that change now...

Reply:
Yep, done that now

Reply:
I had a play around on the Go playground for a potential adapter interface that uses generics. Do you see any problems with an approach like this? Maybe this only works in later versions of Go... 🤔 

https://go.dev/play/p/B87nM2MER7-

Reply:
Oh I'm way behind on generics then! I'm guessing it doesn't work on my local copy as I'm running Go v1.18, but you're right - that does look much better! I think it would need a version bump, but I'm not massively opposed to it.